src/python/WMCore/MicroService/Unified/Common.py

```diff
@@ -139,21 +147,26 @@ def getBlockReplicasAndSize(datasets, phedexUrl, group=None):
     for row in data:
         dataset = row['url'].split('=')[-1]
         if row['data'] is None:
-            print("FAILURE: getBlockReplicasAndSize for %s. Error: %s %s" % (dataset,
-                                                                            row.get('code'),
-                                                                            row.get('error')))
+            print("Failure in getBlockReplicasAndSize for dataset %s. Error: %s %s" % (dataset,
+                                                                                       row.get('code'),
+                                                                                       row.get('error')))
+            dsetBlockSize.setdefault(dataset, None)
             continue
         rows = json.loads(row['data'])
         dsetBlockSize.setdefault(dataset, {})
-        for item in rows['phedex']['block']:
-            block = {item['name']: {'blockSize': item['bytes'], 'locations': []}}
-            for repli in item['replica']:
-                if repli['complete'] == 'y' and repli['subscribed'] == 'y':
-                    if not group:
-                        block[item['name']]['locations'].append(repli['node'])
-                    elif repli['group'] == group:
-                        block[item['name']]['locations'].append(repli['node'])
-            dsetBlockSize[dataset].update(block)
+        try:
+            for item in rows['phedex']['block']:
```

Comment:
Alan,
this block will consume lots of RAM since you have nested dict allocations all over, i.e. dsetBlockSize is a dict which holds block dict which holds yet another (attribute) dicts.

Instead, you can flat it out to simplify the data management. For example, dsetBlockSize can be a flat dict which holds only list of blocks for dataset key. Then you'll have separate flat dicts one to make block-size association and another for block-location one. Therefore instead of one nested structure you'll have 3 flat one. Moreover you can introduce a caching layer for all of them since eventually a dataset and blocks will be fixed. Since you're operating over and over with dataset/blocks your cache will be quite stable. If you want to use this route I suggest that you create separate MSCache class to handle the caches (and again it can be a separate MS micro-service).

Reply:
Valentin, just to make I understood your comment. You suggest to replace the single dict structured like:
```
{"dataset_A": {"blockSize": 123, "locations": [pnn1, pnn2]},
                      {"blockSize": 434, "locations": [pnn2, pnn3]},
 "dataset_B": {...},
},
```

by 3 data structures. First with a list of blocks by dataset:
```
{"dataset_A": [block1, block2],
 "dataset_B": [block3, block5]}
```
then size per block
```
{"block1": 1234,
 "block2": 454}
```
and finally 
```
{"block1": [pnn1, pnn2],
 "block2": [pnn2, pnn3, ...]}
```

This makes us repeat the block name 3 times (well, original + 2 repetitions). While the current approach doesn't repeat it. I totally agree though that it's much cleaner to deal with it in different data structures, but I really don't see how it will save memory consumption.

What we could do is to create a class and constrain it with slots, that should save us some (good?) amount of memory RAM. However, we can't make it generic, thus it would require a different class for each data structure.

For the caching, that's a good point. However, depending on the workflows that we are dealing with, I fear that the memory footprint might be quite high. Keeping a cache around will just make it worse.

Reply:
Alan, yes I'm against nested using dicts, especially when there is NO ANY documentation about their schema.

The idea of using dedicated class with slots to hold block info is actually not to bed. In fact, it will eliminate storing block names 3 times such that you will only need one flat dict {dataset: [BlockObject, BlockObject]}. The BlockObject will hold block name, size and list of sites. The usage of cache is up to you, but I bet having it around may be useful.

Reply:
Check the docstring! ;)
Let me see how much extra work such object will give, if it doesn't make the rest of the logic too hard, I'll push it in.